pruntime/enclave/src/contracts/diem.rs

```diff
@@ -0,0 +1,270 @@
+use crate::std::collections::BTreeMap;
+use crate::std::string::String;
+use crate::std::vec::Vec;
+
+use core::str;
+use parity_scale_codec::{Encode, Decode};
+use serde::{Serialize, Deserialize};
+use sp_core::ecdsa;
+use sp_core::crypto::Pair;
+
+use crate::contracts;
+use crate::TransactionStatus;
+use crate::types::TxRef;
+
+//diem type
+use crate::std::string::ToString;
+use core::convert::TryFrom;
+use diem_types::epoch_change::EpochChangeProof;
+use diem_types::ledger_info::LedgerInfoWithSignatures;
+use diem_types::trusted_state::{TrustedState, TrustedStateChange};
+use diem_types::proof::{
+    AccountStateProof, TransactionInfoWithProof,
+    TransactionAccumulatorProof, SparseMerkleProof,
+};
+use diem_types::account_address::AccountAddress;
+use diem_types::account_state_blob::AccountStateBlob;
+use diem_types::transaction::TransactionInfo;
+use diem_crypto::hash::CryptoHash;
+use diem_types::transaction::{Transaction, SignedTransaction, TransactionPayload};
+use move_core_types::transaction_argument::TransactionArgument;
+
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub struct Amount {
+    pub amount: u64,
+    pub currency: String,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct AccountInfo {
+    pub address: AccountAddress,
+    pub authentication_key: Option<Vec<u8>>,
+    pub sequence_number: u64,
+    pub sent_events_key: String,
+    pub received_events_key: String,
+    pub balances: Vec<Amount>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct TransactionWithProof {
+    transaction_bytes: Vec<u8>,
+
+    epoch_change_proof: EpochChangeProof,
+    ledger_info_with_signatures: LedgerInfoWithSignatures,
+
+    ledger_info_to_transaction_info_proof: TransactionAccumulatorProof,
+    transaction_info: TransactionInfo,
+    transaction_info_to_account_proof: SparseMerkleProof,
+    account_state_blob: AccountStateBlob,
+
+    version: u64,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum Error {
+    NotAuthorized,
+    Other(String),
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum Command {
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum Request {
+    AccountData { account_data_b64: String },
+    VerifyTransaction { account_address: String, transaction_with_proof_b64: String }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum Response {
+    AccountData { size: u32 },
+    VerifyTransaction { total: u32, verified: bool },
+    Error(Error)
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Diem {
+    accounts: Vec<AccountInfo>,
+    transactions: BTreeMap<String, Vec<Transaction>>, //address => Transaction
+    proofs: BTreeMap<String, TransactionWithProof>, //Hash => Proof
+}
```

Comment:
1. We don't need to store (txid => proof). If we want to avoid multiple verification on the same tx, we can just store the txid in a `HashSet`.
2. Theoretically we have to store a `trusted_state` in the Diem contract, representing the latest currently "validated" block, and we update it once we have a new trusted_state. When initialized, we should hard code the genesis trusted_state in it. Then we validate all the transaction based on it, rather than the one come from the query request.
3. Thinking about a way to make sure there's no missing tx when syncing all the historical tx of an account. Maybe we should define a map (account => nonce / sequence) and check the sequence one by one.

Probably better to have an offline meeting to discuss it